From 31d76e8a7e9006b76530f8b6cbb134d198afea2d Mon Sep 17 00:00:00 2001
From: Xufeng Zhang <xufeng.zhang@windriver.com>
Date: Wed, 27 Feb 2013 12:21:53 +0800
Subject: [PATCH] powerpc/booke: Disable debug interrupts if ICMP and IDM are
 enabled

Some applications, like 'usermode-agent', will enable Internal Debug Mode
(IDM) and Instruction Complete Debug Event(ICMP) bits on thread flags.
After On Chip Debugging (eg. MSR_DE bit on Machine State Register)
is enabled on BOOK-E platforms, debug interrupts would be enabled.
This will cause ICMP kernel-mode exception while executing context switch.

To resolve this problem, we don't write DBCR0_ICMP enable bit to SPRN_DBCR0
register on BOOK-E platforms when CONFIG_WR_OCD_DEBUG is enabled.

Signed-off-by: Xufeng Zhang <xufeng.zhang@windriver.com>
---
 arch/powerpc/kernel/process.c |    4 ++++
 1 file changed, 4 insertions(+)

diff --git a/arch/powerpc/kernel/process.c b/arch/powerpc/kernel/process.c
index f95f4fe..7f4f0c0 100644
--- a/arch/powerpc/kernel/process.c
+++ b/arch/powerpc/kernel/process.c
@@ -344,7 +344,11 @@ static void prime_debug_regs(struct thread_struct *thread)
 	mtspr(SPRN_DVC1, thread->dvc1);
 	mtspr(SPRN_DVC2, thread->dvc2);
 #endif
+#if defined (CONFIG_WR_OCD_DEBUG) && defined (CONFIG_BOOKE)
+	mtspr(SPRN_DBCR0, thread->dbcr0 & ~DBCR0_ICMP);
+#else
 	mtspr(SPRN_DBCR0, thread->dbcr0);
+#endif
 	mtspr(SPRN_DBCR1, thread->dbcr1);
 #ifdef CONFIG_BOOKE
 	mtspr(SPRN_DBCR2, thread->dbcr2);
-- 
1.7.9.7

